---
title: GEO database mining
author: 'Bin Xu'
date: '2021-07-17'
slug: geo-database-mining
categories: ["生信"]
tags: ["GEO", "生信技能树"]
---

<script src="{{< blogdown/postref >}}index_files/header-attrs/header-attrs.js"></script>


<p>本笔记来自于<strong>生信技能树</strong>B站视频<a href="https://www.bilibili.com/video/BV1is411H7Hq">《GEO数据库挖掘》</a></p>
<div id="下载表达矩阵geoquery" class="section level2">
<h2>下载表达矩阵(GEOquery)</h2>
<p>本部分笔记来源于我的为知笔记《一文就够-GEO数据库介绍及下载》</p>
<div id="安装包" class="section level3">
<h3>安装包</h3>
<pre class="r"><code>if (!requireNamespace(&quot;BiocManager&quot;, quietly = TRUE))
  install.packages(&quot;BiocManager&quot;)

BiocManager::install(&quot;GEOquery&quot;)</code></pre>
<pre class="r"><code># 加载包
suppressMessages(library(GEOquery))</code></pre>
</div>
<div id="下载表达矩阵" class="section level3">
<h3>下载表达矩阵</h3>
<div id="geoquery函数" class="section level4">
<h4>GEOquery()函数</h4>
<ul>
<li>可以指定<code>GEO =</code>参数以在线下载数据集（GDS、GSE、GSM、GPL均可，但得到的数据对象不同），也可以指定<code>filename =</code>参数读取已经下载的文件，两个参数不能同时指定，只能指定其中一个
<ul>
<li>GSE号返回的对象直接就是expression set对象</li>
<li>指定<code>GEO =</code>和指定<code>filename =</code>得到的对象的数据构成不一样，指定<code>filename =</code>见笔记《读取下载好的表达矩阵(.txt.gz格式)》</li>
</ul></li>
<li><code>destdir</code>参数指定下载文件的存放目录</li>
<li>一些默认参数:
<ul>
<li><code>GSEMatrix = TRUE</code>: 下载Series Matrix files，即表达矩阵</li>
<li><code>AnnotGPL = FALSE</code>: 默认即可</li>
<li><code>getGPL = TRUE</code>: 下载GPL平台的注释信息，You may want to set this to FALSE if you know that you are going to annotate your featureData using Bioconductor tools rather than relying on information provided through NCBI GEO.</li>
</ul></li>
</ul>
<pre class="r"><code>gset &lt;- getGEO(GEO = &quot;GSE42872&quot;, destdir = &quot;.&quot;, getGPL = F)</code></pre>
<pre><code>## Found 1 file(s)</code></pre>
<pre><code>## GSE42872_series_matrix.txt.gz</code></pre>
<pre><code>## Using locally cached version: ./GSE42872_series_matrix.txt.gz</code></pre>
<pre><code>## 
## -- Column specification --------------------------------------------------------
## cols(
##   ID_REF = col_double(),
##   GSM1052615 = col_double(),
##   GSM1052616 = col_double(),
##   GSM1052617 = col_double(),
##   GSM1052618 = col_double(),
##   GSM1052619 = col_double(),
##   GSM1052620 = col_double()
## )</code></pre>
<pre class="r"><code>gset</code></pre>
<pre><code>## $GSE42872_series_matrix.txt.gz
## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 33297 features, 6 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: GSM1052615 GSM1052616 ... GSM1052620 (6 total)
##   varLabels: title geo_accession ... cell type:ch1 (34 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
##   pubMedIds: 24469106 
## Annotation: GPL6244</code></pre>
</div>
<div id="对expression-set对象进行操作的函数" class="section level4">
<h4>对expression set对象进行操作的函数</h4>
<p><code>GEOquery()</code>中给定GSE号返回的对象为expression set</p>
<ul>
<li><code>sampleNames()</code>得到样本名</li>
<li><code>exprs()</code>得到表达矩阵</li>
<li><code>pData()</code>access the phenotypic data (e.g., covariates) and meta-data (e.g., descriptions of covariates) associated with an experiment.</li>
</ul>
<pre class="r"><code>sampleNames(gset)</code></pre>
<pre><code>## [1] &quot;GSM1052615&quot; &quot;GSM1052616&quot; &quot;GSM1052617&quot; &quot;GSM1052618&quot; &quot;GSM1052619&quot;
## [6] &quot;GSM1052620&quot;</code></pre>
<pre class="r"><code># 获得表达矩阵
#  如果一个GSE中有多个平台的数据，那么`gset[[1]]`对应第一个平台的数据，`gset[[2]]`对应第二个平台的数据，以此类推
exprSet &lt;- exprs(gset$GSE42872_series_matrix.txt.gz) # or exprs(gset[[1]])
exprSet[1:6, 1:6]</code></pre>
<pre><code>##         GSM1052615 GSM1052616 GSM1052617 GSM1052618 GSM1052619 GSM1052620
## 7892501    7.24559    6.80686    7.73301    6.18961    7.05335    7.20371
## 7892502    6.82711    6.70157    7.02471    6.20493    6.76554    6.24252
## 7892503    4.39977    4.50781    4.88250    4.36295    4.18137    4.73492
## 7892504    9.48025    9.67952    9.63074    9.69200    9.91324    9.65897
## 7892505    4.54734    4.45247    5.11753    4.87307    5.15505    3.99340
## 7892506    6.80701    6.90597    6.72472    6.77028    6.77058    6.77685</code></pre>
<pre class="r"><code># 获得描述信息
pdata &lt;- pData(gset$GSE42872_series_matrix.txt.gz) # or exprs(gset[[1]])
pdata[1:6, 1:6]</code></pre>
<pre><code>##                                      title geo_accession                status
## GSM1052615     A375 cells 24h Control rep1    GSM1052615 Public on May 20 2014
## GSM1052616     A375 cells 24h Control rep2    GSM1052616 Public on May 20 2014
## GSM1052617     A375 cells 24h Control rep3    GSM1052617 Public on May 20 2014
## GSM1052618 A375 cells 24h Vemurafenib rep1    GSM1052618 Public on May 20 2014
## GSM1052619 A375 cells 24h Vemurafenib rep2    GSM1052619 Public on May 20 2014
## GSM1052620 A375 cells 24h Vemurafenib rep3    GSM1052620 Public on May 20 2014
##            submission_date last_update_date type
## GSM1052615     Dec 12 2012      May 20 2014  RNA
## GSM1052616     Dec 12 2012      May 20 2014  RNA
## GSM1052617     Dec 12 2012      May 20 2014  RNA
## GSM1052618     Dec 12 2012      May 20 2014  RNA
## GSM1052619     Dec 12 2012      May 20 2014  RNA
## GSM1052620     Dec 12 2012      May 20 2014  RNA</code></pre>
</div>
</div>
<div id="下载gpl数据" class="section level3">
<h3>下载GPL数据</h3>
<ul>
<li><code>Meta()</code>函数可以得到描述信息</li>
<li><code>Table()</code>函数可以得到表达矩阵</li>
</ul>
<pre class="r"><code>gpl &lt;- getGEO(&quot;GPL6244&quot;, GSEMatrix = F, destdir = &quot;.&quot;)</code></pre>
<pre><code>## File stored at:</code></pre>
<pre><code>## ./GPL6244.soft</code></pre>
<pre><code>## Warning: 1 parsing failure.
##   row col   expected     actual         file
## 18164  -- 12 columns 11 columns literal data</code></pre>
<pre class="r"><code>names(Meta(gpl)) # 应该没啥用</code></pre>
<pre><code>##  [1] &quot;contact_address&quot;         &quot;contact_city&quot;           
##  [3] &quot;contact_country&quot;         &quot;contact_email&quot;          
##  [5] &quot;contact_institute&quot;       &quot;contact_name&quot;           
##  [7] &quot;contact_phone&quot;           &quot;contact_state&quot;          
##  [9] &quot;contact_web_link&quot;        &quot;contact_zip/postal_code&quot;
## [11] &quot;data_row_count&quot;          &quot;description&quot;            
## [13] &quot;distribution&quot;            &quot;geo_accession&quot;          
## [15] &quot;last_update_date&quot;        &quot;manufacture_protocol&quot;   
## [17] &quot;manufacturer&quot;            &quot;organism&quot;               
## [19] &quot;relation&quot;                &quot;sample_id&quot;              
## [21] &quot;series_id&quot;               &quot;status&quot;                 
## [23] &quot;submission_date&quot;         &quot;taxid&quot;                  
## [25] &quot;technology&quot;              &quot;title&quot;                  
## [27] &quot;web_link&quot;</code></pre>
<pre class="r"><code>gpl_annot &lt;- Table(gpl)

# 不同芯片平台返回的对象需要自行查看列名，然后选择感兴趣信息即可
colnames(gpl_annot)</code></pre>
<pre><code>##  [1] &quot;ID&quot;              &quot;GB_LIST&quot;         &quot;SPOT_ID&quot;         &quot;seqname&quot;        
##  [5] &quot;RANGE_GB&quot;        &quot;RANGE_STRAND&quot;    &quot;RANGE_START&quot;     &quot;RANGE_STOP&quot;     
##  [9] &quot;total_probes&quot;    &quot;gene_assignment&quot; &quot;mrna_assignment&quot; &quot;category&quot;</code></pre>
<pre class="r"><code>gpl_annot[2, c(&quot;ID&quot;, &quot;gene_assignment&quot;, &quot;mrna_assignment&quot;)]</code></pre>
<pre><code>##        ID
## 2 7896738
##                                                                                                                                                                                                                                                                                                                           gene_assignment
## 2 ENST00000328113 // OR4G2P // olfactory receptor, family 4, subfamily G, member 2 pseudogene // --- // --- /// ENST00000492842 // OR4G11P // olfactory receptor, family 4, subfamily G, member 11 pseudogene // --- // --- /// ENST00000588632 // OR4G1P // olfactory receptor, family 4, subfamily G, member 1 pseudogene // --- // ---
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        mrna_assignment
## 2 ENST00000328113 // ENSEMBL // havana:known chromosome:GRCh38:15:101926805:101927707:-1 gene:ENSG00000183909 gene_biotype:unprocessed_pseudogene transcript_biotype:unprocessed_pseudogene // chr1 // 100 // 100 // 31 // 31 // 0 /// ENST00000492842 // ENSEMBL // havana:known chromosome:GRCh38:1:62948:63887:1 gene:ENSG00000240361 gene_biotype:unprocessed_pseudogene transcript_biotype:unprocessed_pseudogene // chr1 // 100 // 100 // 31 // 31 // 0 /// ENST00000588632 // ENSEMBL // havana:known chromosome:GRCh38:19:104535:105471:1 gene:ENSG00000267310 gene_biotype:unprocessed_pseudogene transcript_biotype:unprocessed_pseudogene // chr1 // 100 // 100 // 31 // 31 // 0 /// NONHSAT000016 // NONCODE // Non-coding transcript identified by NONCODE: Linc // chr1 // 100 // 100 // 31 // 31 // 0 /// NONHSAT051704 // NONCODE // Non-coding transcript identified by NONCODE: Linc // chr1 // 100 // 100 // 31 // 31 // 0 /// NONHSAT060106 // NONCODE // Non-coding transcript identified by NONCODE: Linc // chr1 // 100 // 100 // 31 // 31 // 0</code></pre>
</div>
</div>
<div id="读取下载好的表达矩阵.txt.gz格式" class="section level2">
<h2>读取下载好的表达矩阵(.txt.gz格式)</h2>
<div id="方法一geoquery包" class="section level3">
<h3>方法一：GEOquery包</h3>
<p>个人觉得这种方法更方便</p>
<pre class="r"><code>gset2 &lt;- getGEO(filename = &quot;GSE42872_series_matrix.txt.gz&quot;, 
                destdir = &quot;.&quot;, 
                getGPL = F)</code></pre>
<pre><code>## 
## -- Column specification --------------------------------------------------------
## cols(
##   ID_REF = col_double(),
##   GSM1052615 = col_double(),
##   GSM1052616 = col_double(),
##   GSM1052617 = col_double(),
##   GSM1052618 = col_double(),
##   GSM1052619 = col_double(),
##   GSM1052620 = col_double()
## )</code></pre>
<pre class="r"><code>gset2</code></pre>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 33297 features, 6 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: GSM1052615 GSM1052616 ... GSM1052620 (6 total)
##   varLabels: title geo_accession ... cell type:ch1 (34 total)
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
##   pubMedIds: 24469106 
## Annotation: GPL6244</code></pre>
<p>以下如有疑问，详见本笔记的第一部分：<a href="#下载表达矩阵(GEOquery)">下载表达矩阵(GEOquery)</a></p>
<pre class="r"><code>sampleNames(gset2)</code></pre>
<pre><code>## [1] &quot;GSM1052615&quot; &quot;GSM1052616&quot; &quot;GSM1052617&quot; &quot;GSM1052618&quot; &quot;GSM1052619&quot;
## [6] &quot;GSM1052620&quot;</code></pre>
<pre class="r"><code># 获得表达矩阵
exprSet2 &lt;- exprs(gset2)
exprSet2[1:6, 1:6]</code></pre>
<pre><code>##         GSM1052615 GSM1052616 GSM1052617 GSM1052618 GSM1052619 GSM1052620
## 7892501    7.24559    6.80686    7.73301    6.18961    7.05335    7.20371
## 7892502    6.82711    6.70157    7.02471    6.20493    6.76554    6.24252
## 7892503    4.39977    4.50781    4.88250    4.36295    4.18137    4.73492
## 7892504    9.48025    9.67952    9.63074    9.69200    9.91324    9.65897
## 7892505    4.54734    4.45247    5.11753    4.87307    5.15505    3.99340
## 7892506    6.80701    6.90597    6.72472    6.77028    6.77058    6.77685</code></pre>
<pre class="r"><code># 获得描述信息
pdata2 &lt;- pData(gset2)
pdata2[1:6, 1:6]</code></pre>
<pre><code>##                                      title geo_accession                status
## GSM1052615     A375 cells 24h Control rep1    GSM1052615 Public on May 20 2014
## GSM1052616     A375 cells 24h Control rep2    GSM1052616 Public on May 20 2014
## GSM1052617     A375 cells 24h Control rep3    GSM1052617 Public on May 20 2014
## GSM1052618 A375 cells 24h Vemurafenib rep1    GSM1052618 Public on May 20 2014
## GSM1052619 A375 cells 24h Vemurafenib rep2    GSM1052619 Public on May 20 2014
## GSM1052620 A375 cells 24h Vemurafenib rep3    GSM1052620 Public on May 20 2014
##            submission_date last_update_date type
## GSM1052615     Dec 12 2012      May 20 2014  RNA
## GSM1052616     Dec 12 2012      May 20 2014  RNA
## GSM1052617     Dec 12 2012      May 20 2014  RNA
## GSM1052618     Dec 12 2012      May 20 2014  RNA
## GSM1052619     Dec 12 2012      May 20 2014  RNA
## GSM1052620     Dec 12 2012      May 20 2014  RNA</code></pre>
</div>
<div id="方法二read.table函数" class="section level3">
<h3>方法二：read.table()函数</h3>
<p>先打开表达矩阵查看一下数据格式，然后用<code>read.table()</code>函数读取</p>
<p><code>read.table()</code>函数可以直接读取<code>.txt.gz</code>文件，但似乎不能直接用Notepad3打开，打开后乱码，解压后再打开正常显示</p>
<p>该函数常用的参数详见函数帮助</p>
<p>个人觉得相比方法一更繁琐，需要注意的是，表达矩阵读取之后，行名转化为了单独的列</p>
<pre class="r"><code>exprSet3 &lt;- read.table(&quot;GSE42872_series_matrix.txt.gz&quot;, sep = &quot;\t&quot;, comment.char = &quot;!&quot;, header = T)
exprSet3[1:6, 1:6]</code></pre>
<pre><code>##    ID_REF GSM1052615 GSM1052616 GSM1052617 GSM1052618 GSM1052619
## 1 7892501    7.24559    6.80686    7.73301    6.18961    7.05335
## 2 7892502    6.82711    6.70157    7.02471    6.20493    6.76554
## 3 7892503    4.39977    4.50781    4.88250    4.36295    4.18137
## 4 7892504    9.48025    9.67952    9.63074    9.69200    9.91324
## 5 7892505    4.54734    4.45247    5.11753    4.87307    5.15505
## 6 7892506    6.80701    6.90597    6.72472    6.77028    6.77058</code></pre>
<pre class="r"><code>rownames(exprSet3) &lt;- exprSet3[, 1]
exprSet3 &lt;- exprSet3[, -1]
exprSet3[1:6, 1:6]</code></pre>
<pre><code>##         GSM1052615 GSM1052616 GSM1052617 GSM1052618 GSM1052619 GSM1052620
## 7892501    7.24559    6.80686    7.73301    6.18961    7.05335    7.20371
## 7892502    6.82711    6.70157    7.02471    6.20493    6.76554    6.24252
## 7892503    4.39977    4.50781    4.88250    4.36295    4.18137    4.73492
## 7892504    9.48025    9.67952    9.63074    9.69200    9.91324    9.65897
## 7892505    4.54734    4.45247    5.11753    4.87307    5.15505    3.99340
## 7892506    6.80701    6.90597    6.72472    6.77028    6.77058    6.77685</code></pre>
</div>
</div>
<div id="id转换" class="section level2">
<h2>ID转换</h2>
<p>通常会出现多个探针对应同一个基因，取平均值、最大值、中位数等；一个探针也可以对应多个基因</p>
<p>一般有三种方法可以得到芯片探针与gene的对应关系：</p>
<ol style="list-style-type: decimal">
<li>金标准当然是去基因芯片的厂商的官网直接去下载啦！！！</li>
<li>一种是直接用bioconductor的包</li>
<li>一种是从NCBI里面下载文件来解析好！</li>
</ol>
<p>首先，我们说官网，肯定可以找到，不然这种芯片出来就没有意义了！</p>
<p>然后，我们看看NCBI下载的，会比较大：<a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL6947" class="uri">http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL6947</a></p>
<p>这两种方法都比较麻烦，需要一个个的来！</p>
<div id="使用bioconductor中的包" class="section level3">
<h3>使用bioconductor中的包</h3>
<p>常用的芯片平台对应的包可以在<a href="https://www.jianshu.com/p/fb22d4c9ddf9">这里</a>找，现摘录如下（library包时在包名后加上<code>.db</code>）</p>
<p>如果这里找不到，试试Google<code>GPLxxxx R package</code></p>
<pre class="r"><code>&gt;         gpl           organism                   bioc_package
&gt; 
&gt; 1     GPL32       Mus musculus                        mgu74a
&gt; 
&gt; 2     GPL33       Mus musculus                        mgu74b
&gt; 
&gt; 3     GPL34       Mus musculus                        mgu74c
&gt; 
&gt; 6     GPL74       Homo sapiens                        hcg110
&gt; 
&gt; 7     GPL75       Mus musculus                     mu11ksuba
&gt; 
&gt; 8     GPL76       Mus musculus                     mu11ksubb
&gt; 
&gt; 9     GPL77       Mus musculus                     mu19ksuba
&gt; 
&gt; 10    GPL78       Mus musculus                     mu19ksubb
&gt; 
&gt; 11    GPL79       Mus musculus                     mu19ksubc
&gt; 
&gt; 12    GPL80       Homo sapiens                        hu6800
&gt; 
&gt; 13    GPL81       Mus musculus                      mgu74av2
&gt; 
&gt; 14    GPL82       Mus musculus                      mgu74bv2
&gt; 
&gt; 15    GPL83       Mus musculus                      mgu74cv2
&gt; 
&gt; 16    GPL85  Rattus norvegicus                        rgu34a
&gt; 
&gt; 17    GPL86  Rattus norvegicus                        rgu34b
&gt; 
&gt; 18    GPL87  Rattus norvegicus                        rgu34c
&gt; 
&gt; 19    GPL88  Rattus norvegicus                         rnu34
&gt; 
&gt; 20    GPL89  Rattus norvegicus                         rtu34
&gt; 
&gt; 22    GPL91       Homo sapiens                      hgu95av2
&gt; 
&gt; 23    GPL92       Homo sapiens                        hgu95b
&gt; 
&gt; 24    GPL93       Homo sapiens                        hgu95c
&gt; 
&gt; 25    GPL94       Homo sapiens                        hgu95d
&gt; 
&gt; 26    GPL95       Homo sapiens                        hgu95e
&gt; 
&gt; 27    GPL96       Homo sapiens                       hgu133a
&gt; 
&gt; 28    GPL97       Homo sapiens                       hgu133b
&gt; 
&gt; 29    GPL98       Homo sapiens                     hu35ksuba
&gt; 
&gt; 30    GPL99       Homo sapiens                     hu35ksubb
&gt; 
&gt; 31   GPL100       Homo sapiens                     hu35ksubc
&gt; 
&gt; 32   GPL101       Homo sapiens                     hu35ksubd
&gt; 
&gt; 36   GPL201       Homo sapiens                       hgfocus
&gt; 
&gt; 37   GPL339       Mus musculus                       moe430a
&gt; 
&gt; 38   GPL340       Mus musculus                     mouse4302
&gt; 
&gt; 39   GPL341  Rattus norvegicus                       rae230a
&gt; 
&gt; 40   GPL342  Rattus norvegicus                       rae230b
&gt; 
&gt; 41   GPL570       Homo sapiens                   hgu133plus2
&gt; 
&gt; 42   GPL571       Homo sapiens                      hgu133a2
&gt; 
&gt; 43   GPL886       Homo sapiens                     hgug4111a
&gt; 
&gt; 44   GPL887       Homo sapiens                     hgug4110b
&gt; 
&gt; 45  GPL1261       Mus musculus                    mouse430a2
&gt; 
&gt; 49  GPL1352       Homo sapiens                       u133x3p
&gt; 
&gt; 50  GPL1355  Rattus norvegicus                       rat2302
&gt; 
&gt; 51  GPL1708       Homo sapiens                     hgug4112a
&gt; 
&gt; 54  GPL2891       Homo sapiens                       h20kcod
&gt; 
&gt; 55  GPL2898  Rattus norvegicus                     adme16cod
&gt; 
&gt; 60  GPL3921       Homo sapiens                     hthgu133a
&gt; 
&gt; 63  GPL4191       Homo sapiens                       h10kcod
&gt; 
&gt; 64  GPL5689       Homo sapiens                     hgug4100a
&gt; 
&gt; 65  GPL6097       Homo sapiens               illuminaHumanv1
&gt; 
&gt; 66  GPL6102       Homo sapiens               illuminaHumanv2
&gt; 
&gt; 67  GPL6244       Homo sapiens   hugene10sttranscriptcluster
&gt; 
&gt; 68  GPL6947       Homo sapiens               illuminaHumanv3
&gt; 
&gt; 69  GPL8300       Homo sapiens                      hgu95av2
&gt; 
&gt; 70  GPL8490       Homo sapiens   IlluminaHumanMethylation27k
&gt; 
&gt; 71 GPL10558       Homo sapiens               illuminaHumanv4
&gt; 
&gt; 72 GPL11532       Homo sapiens   hugene11sttranscriptcluster
&gt; 
&gt; 73 GPL13497       Homo sapiens         HsAgilentDesign026652
&gt; 
&gt; 74 GPL13534       Homo sapiens  IlluminaHumanMethylation450k
&gt; 
&gt; 75 GPL13667       Homo sapiens                        hgu219
&gt; 
&gt; 76 GPL15380       Homo sapiens      GGHumanMethCancerPanelv1
&gt; 
&gt; 77 GPL15396       Homo sapiens                     hthgu133b
&gt; 
&gt; 78 GPL17897       Homo sapiens                     hthgu133a</code></pre>
<p>读入示例数据，该GSE对应GPL6244</p>
<pre class="r"><code>gset &lt;- getGEO(filename = &quot;GSE42872_series_matrix.txt.gz&quot;, 
               destdir = &quot;.&quot;, 
               getGPL = F)</code></pre>
<pre><code>## 
## -- Column specification --------------------------------------------------------
## cols(
##   ID_REF = col_double(),
##   GSM1052615 = col_double(),
##   GSM1052616 = col_double(),
##   GSM1052617 = col_double(),
##   GSM1052618 = col_double(),
##   GSM1052619 = col_double(),
##   GSM1052620 = col_double()
## )</code></pre>
<pre class="r"><code>exprSet &lt;- exprs(gset)</code></pre>
<pre class="r"><code># 安装包
BiocManager::install(&quot;hugene10sttranscriptcluster.db&quot;)</code></pre>
<pre class="r"><code># 载入包
suppressMessages(library(hugene10sttranscriptcluster.db))</code></pre>
<pre class="r"><code># 通过?hugene10sttranscriptcluster.db
# 查看包里面有那些内容
ls(&quot;package:hugene10sttranscriptcluster.db&quot;)</code></pre>
<pre><code>##  [1] &quot;hugene10sttranscriptcluster&quot;             
##  [2] &quot;hugene10sttranscriptcluster.db&quot;          
##  [3] &quot;hugene10sttranscriptcluster_dbconn&quot;      
##  [4] &quot;hugene10sttranscriptcluster_dbfile&quot;      
##  [5] &quot;hugene10sttranscriptcluster_dbInfo&quot;      
##  [6] &quot;hugene10sttranscriptcluster_dbschema&quot;    
##  [7] &quot;hugene10sttranscriptclusterACCNUM&quot;       
##  [8] &quot;hugene10sttranscriptclusterALIAS2PROBE&quot;  
##  [9] &quot;hugene10sttranscriptclusterCHR&quot;          
## [10] &quot;hugene10sttranscriptclusterCHRLENGTHS&quot;   
## [11] &quot;hugene10sttranscriptclusterCHRLOC&quot;       
## [12] &quot;hugene10sttranscriptclusterCHRLOCEND&quot;    
## [13] &quot;hugene10sttranscriptclusterENSEMBL&quot;      
## [14] &quot;hugene10sttranscriptclusterENSEMBL2PROBE&quot;
## [15] &quot;hugene10sttranscriptclusterENTREZID&quot;     
## [16] &quot;hugene10sttranscriptclusterENZYME&quot;       
## [17] &quot;hugene10sttranscriptclusterENZYME2PROBE&quot; 
## [18] &quot;hugene10sttranscriptclusterGENENAME&quot;     
## [19] &quot;hugene10sttranscriptclusterGO&quot;           
## [20] &quot;hugene10sttranscriptclusterGO2ALLPROBES&quot; 
## [21] &quot;hugene10sttranscriptclusterGO2PROBE&quot;     
## [22] &quot;hugene10sttranscriptclusterMAP&quot;          
## [23] &quot;hugene10sttranscriptclusterMAPCOUNTS&quot;    
## [24] &quot;hugene10sttranscriptclusterOMIM&quot;         
## [25] &quot;hugene10sttranscriptclusterORGANISM&quot;     
## [26] &quot;hugene10sttranscriptclusterORGPKG&quot;       
## [27] &quot;hugene10sttranscriptclusterPATH&quot;         
## [28] &quot;hugene10sttranscriptclusterPATH2PROBE&quot;   
## [29] &quot;hugene10sttranscriptclusterPFAM&quot;         
## [30] &quot;hugene10sttranscriptclusterPMID&quot;         
## [31] &quot;hugene10sttranscriptclusterPMID2PROBE&quot;   
## [32] &quot;hugene10sttranscriptclusterPROSITE&quot;      
## [33] &quot;hugene10sttranscriptclusterREFSEQ&quot;       
## [34] &quot;hugene10sttranscriptclusterSYMBOL&quot;       
## [35] &quot;hugene10sttranscriptclusterUNIPROT&quot;</code></pre>
<pre class="r"><code># 探针与symbol的对应信息
ids &lt;- toTable(hugene10sttranscriptclusterSYMBOL)
head(ids)</code></pre>
<pre><code>##   probe_id     symbol
## 1  7896746   MTND1P23
## 2  7896754 SEPTIN7P13
## 3  7896759  LINC01128
## 4  7896761     SAMD11
## 5  7896779     KLHL17
## 6  7896798    PLEKHN1</code></pre>
<pre class="r"><code>length(unique(ids$symbol)) # 基因个数</code></pre>
<pre><code>## [1] 18859</code></pre>
<pre class="r"><code>tail(sort(table(ids$symbol))) # 查看对应多个探针的基因（探针个数最多的6个）</code></pre>
<pre><code>## 
##    UBTFL1      IGKC    CDK11B    UBE2D3   LRRFIP1 LINC00965 
##         6         7         8         8        10        26</code></pre>
<pre class="r"><code>table(sort(table(ids$symbol))) # 查看相应探针个数的基因各有多少个</code></pre>
<pre><code>## 
##     1     2     3     4     5     6     7     8    10    26 
## 18099   597   131    15     7     5     1     2     1     1</code></pre>
<pre class="r"><code>plot(table(sort(table(ids$symbol)))) # 可视化上一行代码</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<pre class="r"><code># TRUE代表表达矩阵中的探针在package的注释文件中
table(rownames(exprSet) %in% ids$probe_id)</code></pre>
<pre><code>## 
## FALSE  TRUE 
## 13427 19870</code></pre>
<pre class="r"><code>dim(exprSet)</code></pre>
<pre><code>## [1] 33297     6</code></pre>
<pre class="r"><code># 得到包含注释信息的探针的表达矩阵
exprSet &lt;- exprSet[rownames(exprSet) %in% ids$probe_id, ]
dim(exprSet)</code></pre>
<pre><code>## [1] 19870     6</code></pre>
<p><code>match()</code>：匹配两个向量，返回第二个向量在第一个向量匹配位置的下标值</p>
<pre class="r"><code>nrow(exprSet) == nrow(ids)</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code># 使注释文件中的探针顺序与表达矩阵中的探针顺序一致
ids &lt;- ids[match(rownames(exprSet), ids$probe_id), ]

all(rownames(exprSet) == ids$probe_id)</code></pre>
<pre><code>## [1] TRUE</code></pre>
<p>对于一个基因对应多个探针的情况，本例取平均值最大值对应的探针</p>
<p><code>by(data, INDICES, FUN, …, simplify = TRUE)</code>：</p>
<p>将data中的数据，按照INDICES里面的内容拆分成若干个小的data frame，并且在每一小块data frame上应用FUN函数</p>
<pre class="r"><code>probes &lt;- by(exprSet, 
             ids$symbol,
             function(x) rownames(x)[which.max(rowMeans(x))])

class(probes)</code></pre>
<pre><code>## [1] &quot;by&quot;</code></pre>
<pre class="r"><code>probes &lt;- as.character(probes)

exprSet &lt;- exprSet[rownames(exprSet) %in% probes, ]
dim(exprSet)</code></pre>
<pre><code>## [1] 18859     6</code></pre>
<p>在表达矩阵中加入基因名</p>
<pre class="r"><code>dim(exprSet)</code></pre>
<pre><code>## [1] 18859     6</code></pre>
<pre class="r"><code>rownames(exprSet) &lt;- ids[match(rownames(exprSet), ids$probe_id), 2]

dim(exprSet)</code></pre>
<pre><code>## [1] 18859     6</code></pre>
<pre class="r"><code>exprSet[1:2, 1:3]</code></pre>
<pre><code>##            GSM1052615 GSM1052616 GSM1052617
## SEPTIN7P13    8.81485    8.99721    9.19503
## LINC01128     8.75126    8.61650    8.81149</code></pre>
</div>
<div id="如果bioconductor中没有对应的包" class="section level3">
<h3>如果bioconductor中没有对应的包</h3>
<p>需要自行下载GPL注释文件，自行处理再注释</p>
<p>详见本笔记的第一部分：<a href="#下载表达矩阵(GEOquery)">下载表达矩阵(GEOquery)</a></p>
</div>
</div>
<div id="表达矩阵质控" class="section level2">
<h2>表达矩阵质控</h2>
<p>读入示例数据</p>
<p>该部分内容详见本笔记的第三部分：<a href="#ID转换">ID转换</a></p>
<pre class="r"><code>BiocManager::install(&#39;CLL&#39;)</code></pre>
<div id="读入表达矩阵" class="section level3">
<h3>读入表达矩阵</h3>
<pre class="r"><code>suppressPackageStartupMessages(library(CLL))

data(sCLLex)

sCLLex</code></pre>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 12625 features, 22 samples 
##   element names: exprs 
## protocolData: none
## phenoData
##   sampleNames: CLL11.CEL CLL12.CEL ... CLL9.CEL (22 total)
##   varLabels: SampleID Disease
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
## Annotation: hgu95av2</code></pre>
<pre class="r"><code>exprSet &lt;- exprs(sCLLex)

sampleNames(sCLLex)</code></pre>
<pre><code>##  [1] &quot;CLL11.CEL&quot; &quot;CLL12.CEL&quot; &quot;CLL13.CEL&quot; &quot;CLL14.CEL&quot; &quot;CLL15.CEL&quot; &quot;CLL16.CEL&quot;
##  [7] &quot;CLL17.CEL&quot; &quot;CLL18.CEL&quot; &quot;CLL19.CEL&quot; &quot;CLL20.CEL&quot; &quot;CLL21.CEL&quot; &quot;CLL22.CEL&quot;
## [13] &quot;CLL23.CEL&quot; &quot;CLL24.CEL&quot; &quot;CLL2.CEL&quot;  &quot;CLL3.CEL&quot;  &quot;CLL4.CEL&quot;  &quot;CLL5.CEL&quot; 
## [19] &quot;CLL6.CEL&quot;  &quot;CLL7.CEL&quot;  &quot;CLL8.CEL&quot;  &quot;CLL9.CEL&quot;</code></pre>
<pre class="r"><code>pdata &lt;- pData(sCLLex)
head(pdata)</code></pre>
<pre><code>##           SampleID  Disease
## CLL11.CEL    CLL11 progres.
## CLL12.CEL    CLL12   stable
## CLL13.CEL    CLL13 progres.
## CLL14.CEL    CLL14 progres.
## CLL15.CEL    CLL15 progres.
## CLL16.CEL    CLL16 progres.</code></pre>
<pre class="r"><code>group_list &lt;- as.character(pdata[ ,2])

dim(exprSet)</code></pre>
<pre><code>## [1] 12625    22</code></pre>
<pre class="r"><code>exprSet[1:5, 1:5]</code></pre>
<pre><code>##           CLL11.CEL CLL12.CEL CLL13.CEL CLL14.CEL CLL15.CEL
## 1000_at    5.743132  6.219412  5.523328  5.340477  5.229904
## 1001_at    2.285143  2.291229  2.287986  2.295313  2.662170
## 1002_f_at  3.309294  3.318466  3.354423  3.327130  3.365113
## 1003_s_at  1.085264  1.117288  1.084010  1.103217  1.074243
## 1004_at    7.544884  7.671801  7.474025  7.152482  6.902932</code></pre>
</div>
<div id="基因注释" class="section level3">
<h3>基因注释</h3>
<pre class="r"><code>BiocManager::install(&#39;hgu95av2.db&#39;)</code></pre>
<pre class="r"><code>library(hgu95av2.db)</code></pre>
<pre><code>## </code></pre>
<pre class="r"><code>ls(&quot;package:hgu95av2.db&quot;) </code></pre>
<pre><code>##  [1] &quot;hgu95av2&quot;              &quot;hgu95av2.db&quot;           &quot;hgu95av2_dbconn&quot;      
##  [4] &quot;hgu95av2_dbfile&quot;       &quot;hgu95av2_dbInfo&quot;       &quot;hgu95av2_dbschema&quot;    
##  [7] &quot;hgu95av2ACCNUM&quot;        &quot;hgu95av2ALIAS2PROBE&quot;   &quot;hgu95av2CHR&quot;          
## [10] &quot;hgu95av2CHRLENGTHS&quot;    &quot;hgu95av2CHRLOC&quot;        &quot;hgu95av2CHRLOCEND&quot;    
## [13] &quot;hgu95av2ENSEMBL&quot;       &quot;hgu95av2ENSEMBL2PROBE&quot; &quot;hgu95av2ENTREZID&quot;     
## [16] &quot;hgu95av2ENZYME&quot;        &quot;hgu95av2ENZYME2PROBE&quot;  &quot;hgu95av2GENENAME&quot;     
## [19] &quot;hgu95av2GO&quot;            &quot;hgu95av2GO2ALLPROBES&quot;  &quot;hgu95av2GO2PROBE&quot;     
## [22] &quot;hgu95av2MAP&quot;           &quot;hgu95av2MAPCOUNTS&quot;     &quot;hgu95av2OMIM&quot;         
## [25] &quot;hgu95av2ORGANISM&quot;      &quot;hgu95av2ORGPKG&quot;        &quot;hgu95av2PATH&quot;         
## [28] &quot;hgu95av2PATH2PROBE&quot;    &quot;hgu95av2PFAM&quot;          &quot;hgu95av2PMID&quot;         
## [31] &quot;hgu95av2PMID2PROBE&quot;    &quot;hgu95av2PROSITE&quot;       &quot;hgu95av2REFSEQ&quot;       
## [34] &quot;hgu95av2SYMBOL&quot;        &quot;hgu95av2UNIPROT&quot;</code></pre>
<pre class="r"><code>ids &lt;- toTable(hgu95av2SYMBOL)
head(ids)</code></pre>
<pre><code>##    probe_id  symbol
## 1   1000_at   MAPK3
## 2   1001_at    TIE1
## 3 1002_f_at CYP2C19
## 4 1003_s_at   CXCR5
## 5   1004_at   CXCR5
## 6   1005_at   DUSP1</code></pre>
<pre class="r"><code># save(ids, exprSet, pdata, file = &#39;input.Rdata&#39;)

length(unique(ids$symbol))</code></pre>
<pre><code>## [1] 8776</code></pre>
<pre class="r"><code>tail(sort(table(ids$symbol)))</code></pre>
<pre><code>## 
##  RBPMS  TERF1 INPP4A    MYB PTGER3  STAT1 
##      7      7      8      8      8      8</code></pre>
<pre class="r"><code>table(sort(table(ids$symbol)))</code></pre>
<pre><code>## 
##    1    2    3    4    5    6    7    8 
## 6725 1444  451  103   27   16    6    4</code></pre>
<pre class="r"><code>plot(table(sort(table(ids$symbol))))</code></pre>
<p><img src="{{< blogdown/postref >}}index_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<pre class="r"><code>table(rownames(exprSet) %in% ids$probe_id)</code></pre>
<pre><code>## 
## FALSE  TRUE 
##   942 11683</code></pre>
<pre class="r"><code>dim(exprSet)</code></pre>
<pre><code>## [1] 12625    22</code></pre>
<pre class="r"><code>exprSet &lt;- exprSet[rownames(exprSet) %in% ids$probe_id,]
dim(exprSet)</code></pre>
<pre><code>## [1] 11683    22</code></pre>
<pre class="r"><code>ids &lt;- ids[match(rownames(exprSet), ids$probe_id),]
head(ids)</code></pre>
<pre><code>##    probe_id  symbol
## 1   1000_at   MAPK3
## 2   1001_at    TIE1
## 3 1002_f_at CYP2C19
## 4 1003_s_at   CXCR5
## 5   1004_at   CXCR5
## 6   1005_at   DUSP1</code></pre>
<pre class="r"><code>exprSet[1:5,1:5]</code></pre>
<pre><code>##           CLL11.CEL CLL12.CEL CLL13.CEL CLL14.CEL CLL15.CEL
## 1000_at    5.743132  6.219412  5.523328  5.340477  5.229904
## 1001_at    2.285143  2.291229  2.287986  2.295313  2.662170
## 1002_f_at  3.309294  3.318466  3.354423  3.327130  3.365113
## 1003_s_at  1.085264  1.117288  1.084010  1.103217  1.074243
## 1004_at    7.544884  7.671801  7.474025  7.152482  6.902932</code></pre>
<pre class="r"><code>probes &lt;- by(exprSet, ids$symbol,
             function(x) rownames(x)[which.max(rowMeans(x))])
probes&lt;- as.character(probes)

dim(exprSet)</code></pre>
<pre><code>## [1] 11683    22</code></pre>
<pre class="r"><code>exprSet &lt;- exprSet[rownames(exprSet) %in% probes, ]
dim(exprSet)</code></pre>
<pre><code>## [1] 8776   22</code></pre>
<pre class="r"><code>rownames(exprSet) &lt;- ids[match(rownames(exprSet), ids$probe_id), 2]

exprSet[1:5,1:5]</code></pre>
<pre><code>##         CLL11.CEL CLL12.CEL CLL13.CEL CLL14.CEL CLL15.CEL
## MAPK3    5.743132  6.219412  5.523328  5.340477  5.229904
## TIE1     2.285143  2.291229  2.287986  2.295313  2.662170
## CYP2C19  3.309294  3.318466  3.354423  3.327130  3.365113
## CXCR5    7.544884  7.671801  7.474025  7.152482  6.902932
## DUSP1    5.083793  7.610593  7.631311  6.518594  5.059087</code></pre>
<p><a href="http://bio-info-trainee.com/tmp/basic_visualization_for_expression_matrix.html">表达矩阵可视化大全/basic visualization for expression matrix</a></p>
<p>如该网页无法打开，可访问<a href="./basic%20visualization%20for%20expression%20matrix.html">本地文件</a></p>
</div>
</div>
